hw21 Свой cache engine

Taisiya-Boon opened the pull request:
Закончите реализацию MyCache из вебинара.
     Используйте WeakHashMap для хранения значений.

     Добавьте кэширование в DBService из задания про Hibernate ORM или "Самодельный ORM".

     Убедитесь, что ваш кэш действительно работает быстрее СУБД и сбрасывается при недостатке памяти.

Taisiya-Boon committed:
hw21 Свой cache engine
     Закончите реализацию MyCache из вебинара.
     Используйте WeakHashMap для хранения значений.

     Добавьте кэширование в DBService из задания про Hibernate ORM или "Самодельный ORM".

     Убедитесь, что ваш кэш действительно работает быстрее СУБД и сбрасывается при недостатке памяти.

petrelevich commented on hw21-cache/src/main/java/ru/otus/MyCache.java:
```diff
@@ -0,0 +1,57 @@
+    cache.put(key, value);
+    for (WeakReference<HwListener<K, V>> listener : listeners) {
+        if (listener.get() != null) {
+            listener.get().notify(key, value, "put");
```
тут может быть NPE и проверка выше не поможет, т.к. ссылка может протухнуть до второго get().
Надо сначала сделать var ref = listener.get()
потом ref проверить на null и вызвать notify.

petrelevich commented on hw21-cache/src/main/java/ru/otus/MyCache.java:
```diff
@@ -0,0 +1,57 @@
+
+  @Override
+  public void remove(K key) {
+    for (WeakReference<HwListener<K, V>> listener : listeners) {
```
обход лисенеров лучше в отдельный метод вынести.

petrelevich commented on hw21-cache/src/main/java/ru/otus/DbServiceDemo.java:
```diff
@@ -0,0 +1,127 @@
+    }
+
+    private static void withCache() throws Exception {
+        DBServiceUser dbServiceUser = setDbService();
```
Лучше myCache встроить в DBServiceUser, чтобы для пользователя работа с кешем была полностью прозрачной, т.е. незаметной. Пользователь не должен сам к кешу обращаться.

Taisiya-Boon committed:
hw21  Свой cache engine.
      в классе MyCache обход listener вынесен в отдельный метод.
      DbServiceUserWithCacheImpl это реализация DBServiceUser с встроенным MyCache.

Taisiya-Boon committed:
Merge branch 'master' into hw21-cache
